skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
+            ny/2-sup_radius: ny/2+sup_radius] = 1
+    elif len(shape_v) == 3:
+        nx, ny, nz = shape_v
+        sup[nx/2-sup_radius: nx/2+sup_radius,
+            ny/2-sup_radius: ny/2+sup_radius,
+            nz/2-sup_radius: nz/2+sup_radius] = 1
+    return sup
+
+
+def generate_disk_support(sup_radius, shape_v):
+    """
+    Generate support area as a disk for either 2D or 3D cases.
+
+    Parameters
+    ----------
+    sup_radius : float
+        radius of support
+    shape_v : list
+        shape of diffraction pattern, which can be either 2D or 3D case.
+
+    Returns
+    -------
+    sup : array
+        support with a disk area
+    """
+    sup = np.zeros(shape_v)
+    dummy = _dist(shape_v)
+    sup_index = np.where(dummy <= sup_radius)
+    sup[sup_index] = 1
+    return sup
+
+
+def cdi_recon(diff_array, sample_obj, sup,
+              beta=1.15, start_ave=0.8, pi_modulus_flag='Complex',
+              sw_flag=True, sw_sigma=0.5, sw_threshold=0.1, sw_start=0.2,
+              sw_end=0.8, sw_step=10, n_iterations=1000):
+    """
+    Run reconstruction with difference map algorithm.
+
+    Parameters
+    ---------
+    diff_array : array
+        diffraction pattern from experiments
+    sample_obj : array
+        initial sample with phase
+    sup : array
+        initial support
+    beta : float, optional
+        feedback parameter for difference map algorithm.
+        default is 1.15.
+    start_ave : float, optional
+        define the point to start doing average.
+        default is 0.8.
+    pi_modulus_flag : str, optional
+        'Complex' or 'Real', defining the way to perform pi_modulus calculation.
+        default is 'Complex'.
+    sw_flag : Bool, optional
+        flag to use shrinkwrap algorithm or not.
+        default is True.
+    sw_sigma : float, optional
+        gaussian width used in sw algorithm.
+        default is 0.5.
+    sw_threshold : float, optional
+        shreshold cut in sw algorithm.
+        default is 0.1.
+    sw_start : float, optional
+        at which point to start to do shrinkwrap.
+        defualt is 0.2
+    sw_end : float, optional
+        at which point to stop shrinkwrap.
+        defualt is 0.8
+    sw_step : float, optional
+        the frequency to perform sw algorithm.
+        defualt is 10
+    n_iterations : int, optional
+        number of iterations to run.
+        default is 1000.
+
+    Returns
+    -------
+    obj_ave : array
+        reconstructed sample object
+    error_dict : dict
+        Error information for all iterations. The dict keys include
+        obj_error, diff_error and sup_error. Obj_error is a list of
+        the relative error of sample object. Diff_error is calculated as
+        the difference between new diffraction pattern and the original
+        diffraction pattern. And sup_error stores the size of the
+        sample support.
+    """
+
+    diff_array = np.array(diff_array)     # diffraction data
```

Comment:
I think you might get a bit confused about the algorithm. diffraction array has nothing to do with fft or ifft. 

fft/ifft only works on sample_obj within the iteration part below. I didn't see the reason why it is better to take one fftshift out of the helper function, while keep the second fftshift inside the function. This will make the code difficult to read.